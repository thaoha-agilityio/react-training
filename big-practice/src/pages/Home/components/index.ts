export * from './Hero';
export * from './BrowTheRange';
export * from './Adverse';
export * from './OurProducts';
