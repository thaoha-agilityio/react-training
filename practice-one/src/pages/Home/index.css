.main-content {
  flex: 1;
  box-shadow: 0px 2px 10px var(--color-primary-250);
  border-radius: 6px;
  max-height: 1000px;
  overflow: auto;
}

.page-wrapper {
  display: flex;
  gap: 20px;
  justify-content: end;
  padding: 24px;
}

.content {
  min-height: calc(100vh - 218px);
  box-shadow: 0px 2px 10px var(--color-primary-250);
  padding: 24px;
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.dropdown {
  padding: 20px 20px 24px;
  box-shadow: 0px 2px 10px var(--color-primary-200);
  border-radius: 10px;
}

.dropdown h2 {
  font-family: 'Inter';
  font-weight: 600;
  font-size: 25px;
  line-height: calc(32 / 25);
  color: var(--color-primary);
}

.dropdown-menu {
  margin-top: 20px;
  display: flex;
  justify-content: space-between;
}
