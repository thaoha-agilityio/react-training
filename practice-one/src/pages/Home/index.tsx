import React from 'react';

// Components
import Header from '../../components/Header';
import Footer from '../../components/Footer';
import Content from '../../components/Content';
import SearchFilter from '../../components/SearchFilter';
import DropdownMenu from '../../components/DropdownMenu';
import TextField from '../../components/TextField';
import Button from '../../components/Button';
import UserList from '../../components/Table/UserList';

import { OPTIONS_ROLE, OPTIONS_PROJECT } from '../../constants/dropdown';
import { avatars, userNames, users, emails } from '../../mocks/users';
import { IUser } from '../../types/IUser';
import { createID } from '../../helpers/createId';
import { random } from '../../helpers/random';

// CSS
import './index.css';

interface IProps {
  children?: React.ReactNode;
}

interface IState {
  value: string;
  userList: IUser[];
  selectedFilter: IUser[];
}

class Home extends React.Component<IProps, IState> {
  state = {
    userList: users,
    value: '',
    selectedFilter: [],
  };

  // Add a user in data
  handleAddUser = (): void => {
    const newUser = {
      id: createID(),
      name: random(userNames),
      email: random(emails),
      avatar: random(avatars),
    } as IUser;

    this.setState({ userList: [...this.state.userList, newUser] });
  };

  // Delete by Id
  handleDeleteUser = (id: string): void => {
    const currentUserList = this.state.userList.filter((item) => item.id !== id);

    this.setState({ userList: currentUserList });
  };

  // Search by name
  handleSearchUser = (): void => {
    const { userList } = this.state;

    const searchList = userList.filter(
      (item: IUser) => item.name.toLowerCase().search(this.state.value.toLowerCase()) >= 0
    );

    this.setState({ userList: searchList });
  };

  // Get value input
  handleChangeInput = (event: React.FormEvent<HTMLInputElement>): string => {
    this.setState({ value: event.currentTarget.value });

    return this.state.value;
  };

  // Handle filter user by role
  handleFilterUserByRole = (event: React.ChangeEvent<HTMLSelectElement>): IUser[] => {
    const { userList } = this.state;

    // get project name from value dropdown
    const valueSelected = event.target.value;

    const usersFilter = userList.filter((item: IUser) =>
      item.projects?.some((value) => value.role === valueSelected)
    );

    this.setState({ selectedFilter: usersFilter });

    return usersFilter;
  };

  // Handle filter user by role & project name
  handleFilterUserByProject = (event: React.ChangeEvent<HTMLSelectElement>): IUser[] => {
    const { selectedFilter } = this.state;
    const values = event.target.value;

    const usersFilter = selectedFilter.filter((item: IUser) =>
      item.projects?.some((value) => value.projectName === values)
    );

    this.setState({ userList: usersFilter });

    return this.state.userList;
  };

  render(): React.ReactNode {
    const { value, userList } = this.state;

    return (
      <div className="container">
        <Header />
        <Content>
          <SearchFilter>
            <DropdownMenu
              options={OPTIONS_ROLE}
              size="medium"
              onChange={this.handleFilterUserByRole}
            />
            <DropdownMenu
              options={OPTIONS_PROJECT}
              size="medium"
              onChange={this.handleFilterUserByProject}
            />
          </SearchFilter>
          <div className="main-content">
            <div className="page-wrapper">
              <TextField
                type="text"
                onChange={this.handleChangeInput}
                onClick={this.handleSearchUser}
                value={value}
              />
              <Button variant="primary" size="medium" onClick={this.handleAddUser}>
                Add User
              </Button>
            </div>
            <UserList userList={userList} onDelete={this.handleDeleteUser} />
          </div>
        </Content>
        <Footer />
      </div>
    );
  }
}

export default Home;
