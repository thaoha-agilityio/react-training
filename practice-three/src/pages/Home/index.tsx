import React, {
  ChangeEvent,
  lazy,
  useCallback,
  useEffect,
  useState,
} from "react";
import { ThemeProvider } from "styled-components";

// Custom hooks
import { useCategories, useDebounce } from "../../hooks";
import { useBooks } from "../../hooks/useBooks";

// Components
import Header from "../../components/Header";
import SubHeader from "./components/SubHeader";
import Books from "./components/Books";
import SideBar from "./components/SideBar";

// Themes
import { darkTheme, lightTheme } from "../../themes";
import { Container } from "../../styled-common";

// Constant
import { KEY_NAME_ESC } from "../../constants/actions";

// Styled
import { MainContentStyled } from "./index.styled";

// React lazy
const DetailModal = lazy(() => import("../../components/Modal/DetailModal"));
const FilterModal = lazy(() => import("../../components/Modal/FilterModal"));

const Home = (): React.ReactElement => {
  const { searchBooks, getBookById, isGridView } = useBooks();

  const { categories } = useCategories();

  const [isDarkTheme, setIsDarkTheme] = useState(false);

  const [inputValue, setInputValue] = useState<string>("");
  const debounceValue = useDebounce(inputValue);

  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const [selectedBookId, setSelectedBookId] = useState<string>("");
  const [isModalFilterOpen, setIsModalFilterOpen] = useState<boolean>(false);

  // Change dark-light mode
  const handleToggleTheme = useCallback(() => {
    setIsDarkTheme(!isDarkTheme);
  }, [isDarkTheme]);

  // Get value input
  const handleChangeInput = useCallback(
    async (event: ChangeEvent<HTMLInputElement>): Promise<void> => {
      const value = event.target.value;

      setInputValue(value);

      // If remove all text then fetch data
      if (value === "") {
        await searchBooks(value);
      }
    },
    []
  );

  useEffect(() => {
    // Declare the data fetching function
    const fetchBooks = async () => {
      await searchBooks(debounceValue);
    };

    if (debounceValue) {
      fetchBooks();
    }
  }, [debounceValue]); // Only call effect if debounced debounceValue changes

  // Set id when click item
  const handleSetSelectedBookId = useCallback((id: string) => {
    setSelectedBookId(id);
    handleToggleModal();
  }, []);

  const handleToggleModal = useCallback((): void => {
    setIsModalOpen((prev) => !prev);
  }, [isModalOpen]);

  // Close detail modal by keyboard
  const handleCloseByKeyboard = useCallback((event: KeyboardEvent): void => {
    if (event.keyCode === KEY_NAME_ESC) {
      handleToggleModal();
    }
  }, []);

  // Handle show/close filter component
  const handleToggleFilterModal = useCallback(() => {
    setIsModalFilterOpen((prev) => !prev);
  }, []);

  return (
    <ThemeProvider theme={isDarkTheme ? darkTheme : lightTheme}>
      <Container>
        <Header
          isDarkTheme={isDarkTheme}
          onToggleTheme={handleToggleTheme}
          onChange={handleChangeInput}
        />
        <SubHeader onToggleFilterModal={handleToggleFilterModal} />
        <MainContentStyled>
          <SideBar categories={categories} />
          <Books
            onSetSelectedBookId={handleSetSelectedBookId}
            isGridView={isGridView}
          />
        </MainContentStyled>
      </Container>
      {isModalOpen && (
        <DetailModal
          onCloseModal={handleToggleModal}
          book={getBookById(selectedBookId)}
          onCloseByKeyboard={handleCloseByKeyboard}
          isDarkTheme={isDarkTheme}
          onToggleTheme={handleToggleTheme}
        />
      )}
      {isModalFilterOpen && (
        <FilterModal
          width={268}
          height={331}
          top={20}
          right={70}
          onToggleFilterModal={handleToggleFilterModal}
        />
      )}
    </ThemeProvider>
  );
};

export default Home;
