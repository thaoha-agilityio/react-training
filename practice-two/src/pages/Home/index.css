.home {
  width: 1259px;
  border-radius: 10px;
  margin: auto;
  border: 1px solid;
}
