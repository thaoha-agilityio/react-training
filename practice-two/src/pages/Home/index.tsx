import { ChangeEvent, useCallback, useContext, useState } from 'react';

import Header from '@/components/Header';
import SubHeader from './SubHeader';
import Books from './Books.tsx';
import SideBar from '@/pages/Home/SideBar';
import DetailModal from '@/components/Modal/DetailModal';
import FilterModal from '@/components/Modal/FilterModal';

import { BooksContext } from '@/contexts/BooksContext';
import { CategoriesContext } from '@/contexts/CategoriesContext';
import { ThemeContext } from '@/contexts/ThemeContext ';

import './index.css';

const Home = (): JSX.Element => {
  const { ids, searchBooks, getBookById, isGridView } = useContext(BooksContext);
  const { categories } = useContext(CategoriesContext);
  const { isDarkMode } = useContext(ThemeContext);

  const [valueInput, setValueInput] = useState<string>('');
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
  const [selectedBookId, setSelectedBookId] = useState<string>('');
  const [isModalFilterOpen, setIsModalFilterOpen] = useState<boolean>(false);

  // Handle Search books
  const handleChangeInput = useCallback(
    async (event: ChangeEvent<HTMLInputElement>): Promise<void> => {
      // Get value input
      const values = event.target.value;

      // Search books by name
      await searchBooks(values);

      setValueInput(values);
    },
    [valueInput]
  );

  // Handle show modal
  const handleShowModal = useCallback(
    (id: string): void => {
      setIsModalOpen(true);

      // Get id when click item
      setSelectedBookId(id);
    },
    [isModalOpen]
  );

  // Handle close modal
  const handleCloseModal = (): void => {
    setIsModalOpen(false);
  };

  // Handle show filter modal
  const handleShowFilterModal = useCallback((): void => {
    setIsModalFilterOpen(true);
  }, [isModalFilterOpen]);

  // Handle close modal
  const handleCloseFilterModal = (): void => {
    setIsModalFilterOpen(false);
  };

  return (
    <div className={`home ${isDarkMode ? 'dark-theme' : 'light-theme'}  `}>
      <div className="container">
        <Header onChange={handleChangeInput} />
        <SubHeader onShowModal={handleShowFilterModal} />
        <div className="contents">
          <SideBar categories={categories} />
          <Books ids={ids} onShowModal={handleShowModal} isGridView={isGridView} />
        </div>
        {isModalOpen && (
          <DetailModal onCloseModal={handleCloseModal} book={getBookById(selectedBookId)} />
        )}

        {isModalFilterOpen && <FilterModal onCloseModal={handleCloseFilterModal} />}
      </div>
    </div>
  );
};

export default Home;
