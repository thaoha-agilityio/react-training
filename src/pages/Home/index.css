.main-content {
  flex: 1;
  box-shadow: 0px 2px 5px var(--color-primary);
  border-radius: 5px;
  max-height: 1000px;
  overflow: auto;
}

.page-wrapper {
  display: flex;
  gap: 20px;
  justify-content: end;
  padding: 24px;
}
