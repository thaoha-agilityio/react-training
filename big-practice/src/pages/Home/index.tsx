import Container from '@components/Container';
import { Hero, BrowTheRange, Adverse, OurProducts } from './components';

const Home = () => {
  return (
    <>
      <Hero />
      <Container>
        <BrowTheRange />
        <OurProducts />
        <Adverse />
      </Container>
    </>
  );
};

export default Home;
