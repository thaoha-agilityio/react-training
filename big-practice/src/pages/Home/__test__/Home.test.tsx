import { renderWithRouterAndQuery } from '@helpers';
import Home from '../index';

describe('Home page', () => {
  it('should render component correctly', () => {
    const { container } = renderWithRouterAndQuery(<Home />);

    expect(container).toBeInTheDocument();
  });
});
