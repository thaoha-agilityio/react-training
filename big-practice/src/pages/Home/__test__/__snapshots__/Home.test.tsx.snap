// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should render component correctly 1`] = `
<div>
  <section
    class="css-1x1cs40"
  >
    <div
      class="css-1rrw7j4"
    />
    <div
      class="css-1nio0lb"
    >
      <div
        class="css-1vvasvl"
      >
        <p
          class="chakra-text css-b2hpjo"
        >
          New Arrival
        </p>
        <h2
          class="chakra-heading css-sbfz88"
        >
          discover our new collection
        </h2>
        <p
          class="chakra-text css-152l52r"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut elit tellus, luctus nec ullamcorper mattis.
        </p>
        <button
          class="chakra-button css-1gvc6wf"
          type="button"
        >
          buy now
        </button>
      </div>
    </div>
  </section>
  <div
    class="chakra-container css-0"
  >
    <section
      class="css-0"
    >
      <div
        class="css-1h1egd6"
      >
        <h3
          class="chakra-heading css-hvgyr9"
        >
          Browse The Range
        </h3>
        <p
          class="chakra-text css-lzwicj"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
      </div>
      <div
        class="css-1vjvxz0"
      >
        <div
          class="chakra-stack css-1x72rqo"
        >
          <img
            alt="dining"
            class="chakra-image css-3qdny7"
            src=""
          />
          <p
            class="chakra-text css-0"
          >
            Dining
          </p>
        </div>
        <div
          class="chakra-stack css-1x72rqo"
        >
          <img
            alt="bedroom"
            class="chakra-image css-3qdny7"
            src=""
          />
          <p
            class="chakra-text css-0"
          >
            Bedroom
          </p>
        </div>
        <div
          class="chakra-stack css-1x72rqo"
        >
          <img
            alt="living"
            class="chakra-image css-3qdny7"
            src=""
          />
          <p
            class="chakra-text css-0"
          >
            Living
          </p>
        </div>
      </div>
    </section>
    <section
      class="chakra-stack css-yet0nr"
    >
      <h3
        class="chakra-heading css-11yk9zt"
      >
        Our Products
      </h3>
      <div
        class="css-sipp05"
      >
        <div
          class="chakra-spinner css-1y7joxr"
        >
          <span
            class="css-8b45rq"
          >
            Loading...
          </span>
        </div>
      </div>
    </section>
    <section
      class="css-175tbkm"
    >
      <div
        class="css-1l4w6pd"
      >
        <h3
          class="chakra-heading css-15ca6y6"
        >
          #FuniroFurniture
        </h3>
      </div>
      <img
        alt="adverse"
        class="chakra-image css-uwwqev"
        src=""
      />
    </section>
  </div>
</div>
`;
